### PR TITLE
Fields modifications including trait statistics graphs

### LINKED ISSUE #52: Traits sorting: get back the original order after sorting the traits
It would be nice to add the possibility to get the traits in the original order (traits order in the original file after importing) after sorting them.  The option could be added in the sorting menu.

### LINKED ISSUE #76: Replace Camera trait with built-in camera2 trait

### LINKED ISSUE #113: Basic statistics about fields
Show basic stats about imported fields: trait averages, frequencies, etc.

### LINKED ISSUE #178: Modify Fields page and workflow
- [x] Add beta setting to use new fields workflow

### Fields list

- [x] #805
- [x] #775
- [x] Most recently edited or imported file is pinned to the top of list
- [x] Long press (or press icon) to enter selector mode that allows multiple list items to be selected
- [x] Selector mode should have specific actions on toolbar: delete, export, archive
- [x] #267
- [x] When in selector mode, add a select all button to the toolbar (double checks)
- [x] Move add field button to FAB
- [x] Add "new field" item to add field dialog and remove from toolbar
- [x] Sort field list by import source, **import date**, edit date, export date, or **name**.
- [x] highlight current sort option and add reset
- [x] add shaded circle to show % complete for trait detail sections
- [x] include icons in add field option list and move create new to end

### Per field activity

- [x] Move existing information to per field activity
- [x] Toolbar labeled 'Field Detail' with rename, sort, and delete actions
- [x] Collapsible design card that shows field details - name+import date and source, entry number and sort order, last collected by person
- [x] Click action cards for collect with last edit date, Export with last export date, Sync (if brapi) with last sync date
- [x] Trait card with num traits, num observations, and recyclerview items for each collected traits with type icon, observation count
- [x] Format dates to be semantic (i.e. Today, Yesterday, Never, Last year, Mon DD)
- [x] Ignore data from deleted traits
- [x] Add trait statistics graphs - histograms for numeric data
- [x]  Add trait statistics graphs - bar charts for ordinal/categorical data
- [x] https://github.com/PhenoApps/Field-Book/issues/960

### LINKED ISSUE #197: Add local search/filtering to brapi results

### LINKED ISSUE #267: Bulk export of files

### LINKED ISSUE #268: Rename Field
Edit the field_alias and ensure it is used across layouts

### LINKED ISSUE #805: Update field list to show number of entries and use recyclerview

### LINKED ISSUE #846: [BUG]: FieldEditorActivity's Sync list item crashes app
### Description

Sync crashes app if a url is not configured

### Expected behavior

_No response_

### Field Book version

_No response_

### LINKED ISSUE #875: BrAPI Fields List Overhaul
- [x] Modify pagination to scrolling recylcerview- going to the end loads the next page and appends it to the list
- [x] Add search to top toolbar
- [x] Move load fields to top toolbar as an icon
- [x] Combine Filter options to a dialog
- [x] Add year as a filter criteria
- [x] Save/load filters
- [x] Add each filter criteria as a sub-top toolbar with an X to delete that criteria
- [x] Stylize server name (via serverinfo) and add as a sub-top toolbar
- [x] Replace preview field with a bottom banner that only displays once a field has been selected
- [x] Update import dialog to alertdialog style
- [x] Rename options in import dialog to "Cancel" and "Import"
- [x] Move observation level to after field is selected

### LINKED ISSUE #876: BrAPI Trait List Overhaul
- [x] Modify pagination to scrolling recylcerview- going to the end loads the next page and appends it to the list
- [x] Add search to top toolbar
- [x] Move load traits to top toolbar as an icon
- [x] Save/load searches on activityStart
- [x] Searches added as a sub-top toolbar with an X to delete that criteria
- [x] Need to keep a running list of traits to import if multiple searches are performed/cleared
- [x] Stylize server name (via serverinfo) and add as a sub-top toolbar
- [x] Add total number of traits in current list (if this is gettable) under servername. Update this number as different criteria are applied
- [x] Add BrAPI Trait Import to top toolbar
- [x] Replace `Save Traits` with a bottom banner that only displays once a trait has been selected. Banner should include number of traits 
- [x] Include the trait format in the trait listitems
- [x] Add support for filters: species
- [x] Filters should be added as a sub-top toolbar with an X to delete/reload the returned list

### LINKED ISSUE #883: GoPro Hero 12 Compatibility

### LINKED ISSUE #889: [BUG]: USB cameras record photos in low resolution.
### Description

There is a problem with the quality of image recording on USB cameras.

I tried to connect 3 different USB cameras, on 2 different devices.

When saving the photo with the application (USB Camera Trait), they are saved in a low resolution. For a single camera tested on 2 Android devices, the recording resolution is different. This could mean that it is not the native image from the camera that is recorded but the preview on the screen of the Android device.

For example, when recording photos with a 4K camera and a 13MP with Fielbook USB Camera Trait, I obtain an image around 1000*800pixels and 1.5Mb.

### Expected behavior

The application should either allow you to save the photo in the highest definition of the camera, or allow you to select from several resolutions on the camera.

### Field Book version

5.6.5

### LINKED ISSUE #960: Theme and layout updates
- [x] Fields multiple export needs to implement theming
- [x] Alpha fields names are not legible with large font/tablets
- [x] Center trait formats in new trait dialog
- [x] Some activities are black when device is using system dark mode
- [x] Config activity: listview to bottom of parent instead of top of FAB

### LINKED ISSUE #968: Trait sort by name should ignore case

### LINKED ISSUE #998: [BUG]: Photo trait doesn't work if trait name contains "|" or ":"
### Description

A trait name such as  "fruit plot harvest picture [photo]|UNH_01:0001025" won't save a photo unless the name is changed to exclude both the ":" and "|". Unfortunately, both of those characters are standard for trait imports from Breedbase due to the ontology format they use. 
 

### Expected behavior

_No response_

### Field Book version

_No response_

### PR DESCRIPTION
**Pull request by bellerbrock**
## Description

_Provide a summary of your changes including motivation and context.
If these changes fix a bug or resolves a feature request, be sure to link to that issue._

Closes #178
Closes #968
Closes #960
Closes #52 
Closes #113 
Closes #76 
Closes #268 
Closes #267 
Closes #846 
Closes #805 
Closes #889 
Closes #883 
Closes #998 
Closes #197 
Closes #875 
Closes #876

## Type of change

_What type of changes does your code introduce? Put an `x` in boxes that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [x] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated relevant documentation

## Changelog entry

_Please add a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Added trait statistics graphs to field detail page
```

### CODE REVIEW

**Review by chaneylc — CHANGES_REQUESTED** [review]
This is looking really awesome, great work! I've made some minor comments and one semi-major comment. I understand some of my comments might be a limitation of the library. For example of creating a Chartable interface using the Formats class, check out the offbeat camera PR and look at coders and value presenters.

File: app/src/main/java/com/fieldbook/tracker/charts/HistogramChartHelper.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        chart.axisRight.isEnabled = false
+        chart.setScaleEnabled(false)
+        chart.setDragEnabled(false)
+        chart.setHighlightPerTapEnabled(false)
+        chart.legend.isEnabled = false
+
+        val description = Description()
+        description.text = ""
+        chart.description = description
+        chart.setNoDataText("No data available")
```
[the remark this answers is not in the document]
Comment (bellerbrock) [reply app/src/main/java/com/fieldbook/tracker/charts/HistogramChartHelper.kt]: [147ad13](https://github.com/PhenoApps/Field-Book/pull/985/commits/147ad1397086ab35dd6f82f5d60ea9dded37c650)

File: app/src/main/java/com/fieldbook/tracker/charts/HistogramChartHelper.kt [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+//        if (range.compareTo(BigDecimal.ZERO) == 0) {
+//            chart.visibility = View.GONE
+//            return
+//        }
+
+        chart.visibility = View.VISIBLE
+
+        val distinctValuesCount = observations.distinct().size
+        val binCount = minOf(10, distinctValuesCount)
+        val binSize = range.divide(BigDecimal(binCount), BigDecimal.ROUND_UP)
```
[the remark this answers is not in the document]
Comment (bellerbrock) [reply app/src/main/java/com/fieldbook/tracker/charts/HistogramChartHelper.kt]: [147ad13](https://github.com/PhenoApps/Field-Book/pull/985/commits/147ad1397086ab35dd6f82f5d60ea9dded37c650)

File: app/src/main/java/com/fieldbook/tracker/charts/HistogramChartHelper.kt [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            binnedObservations[binIndex] = (binnedObservations[binIndex] ?: 0) + 1
+        }
+
+        val entries = binnedObservations.map { (binIndex, count) ->
+            val binStart = minValue.add(binSize.multiply(BigDecimal(binIndex)))
+            val binCenter = binStart.add(binSize.divide(BigDecimal(2)))
+            BarEntry(binCenter.toFloat(), count.toFloat())
+        }
+
+        val dataSet = BarDataSet(entries, "Observations")
```
[the remark this answers is not in the document]
Comment (bellerbrock) [reply app/src/main/java/com/fieldbook/tracker/charts/HistogramChartHelper.kt]: [147ad13](https://github.com/PhenoApps/Field-Book/pull/985/commits/147ad1397086ab35dd6f82f5d60ea9dded37c650)

File: app/src/main/java/com/fieldbook/tracker/charts/BarChartHelper.kt [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        chart.axisRight.isEnabled = false
+        chart.setScaleEnabled(false)
+        chart.setDragEnabled(false)
+        chart.setHighlightPerTapEnabled(false)
+        chart.legend.isEnabled = false
+
+        val description = Description()
+        description.text = ""
+        chart.description = description
+        chart.setNoDataText("No data available")
```
[the remark this answers is not in the document]
Comment (bellerbrock) [reply app/src/main/java/com/fieldbook/tracker/charts/BarChartHelper.kt]: [147ad13](https://github.com/PhenoApps/Field-Book/pull/985/commits/147ad1397086ab35dd6f82f5d60ea9dded37c650)

File: app/src/main/java/com/fieldbook/tracker/charts/HistogramChartHelper.kt [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+//        }
+
+        chart.visibility = View.VISIBLE
+
+        val distinctValuesCount = observations.distinct().size
+        val binCount = minOf(10, distinctValuesCount)
+        val binSize = range.divide(BigDecimal(binCount), BigDecimal.ROUND_UP)
+
+        val binnedObservations = mutableMapOf<Int, Int>()
+        for (observation in observations) {
```
[the remark this answers is not in the document]
Comment (bellerbrock) [reply app/src/main/java/com/fieldbook/tracker/charts/HistogramChartHelper.kt:54]: I tested a bit but couldn't find exactValueInt(), you mean toIntExact()? In any case yeah i think toInt() has been fine precision in all the testing I've done.

File: app/src/main/java/com/fieldbook/tracker/charts/HistogramChartHelper.kt [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+import com.github.mikephil.charting.components.XAxis
+import com.github.mikephil.charting.components.YAxis
+import com.github.mikephil.charting.data.BarData
+import com.github.mikephil.charting.data.BarDataSet
+import com.github.mikephil.charting.data.BarEntry
+import com.github.mikephil.charting.formatter.ValueFormatter
+import java.math.BigDecimal
+import kotlin.math.ceil
+
+object HistogramChartHelper {
```
[the remark this answers is not in the document]
Comment (bellerbrock) [reply app/src/main/java/com/fieldbook/tracker/charts/HistogramChartHelper.kt:22]: fixed [4988313](https://github.com/PhenoApps/Field-Book/pull/985/commits/4988313e6586dac089d898213c407be5264b8b13)

File: app/src/main/java/com/fieldbook/tracker/charts/BarChartHelper.kt [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import com.github.mikephil.charting.components.Description
+import com.github.mikephil.charting.components.XAxis
+import com.github.mikephil.charting.components.YAxis
+import com.github.mikephil.charting.data.BarData
+import com.github.mikephil.charting.data.BarDataSet
+import com.github.mikephil.charting.data.BarEntry
+import com.github.mikephil.charting.formatter.ValueFormatter
+import kotlin.math.ceil
+
+object BarChartHelper {
```
[the remark this answers is not in the document]
Comment (bellerbrock) [reply app/src/main/java/com/fieldbook/tracker/charts/BarChartHelper.kt]: fixed with [fbf7116](https://github.com/PhenoApps/Field-Book/pull/985/commits/fbf711698298bbc0bf22a711b0e8b61abcedc45f) and [0897a99](https://github.com/PhenoApps/Field-Book/pull/985/commits/0897a99f3e600089b1f0396111454c08c020f3f9)

File: app/src/main/java/com/fieldbook/tracker/charts/HistogramChartHelper.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (bellerbrock) [reply app/src/main/java/com/fieldbook/tracker/charts/HistogramChartHelper.kt:22]: fixed [4988313](https://github.com/PhenoApps/Field-Book/pull/985/commits/4988313e6586dac089d898213c407be5264b8b13)

File: app/src/main/java/com/fieldbook/tracker/charts/HistogramChartHelper.kt [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+//        if (range.compareTo(BigDecimal.ZERO) == 0) {
+//            chart.visibility = View.GONE
+//            return
+//        }
+
+        chart.visibility = View.VISIBLE
+
+        val distinctValuesCount = observations.distinct().size
+        val binCount = minOf(10, distinctValuesCount)
```
[the remark this answers is not in the document]
Comment (bellerbrock) [reply app/src/main/java/com/fieldbook/tracker/charts/HistogramChartHelper.kt]: 10 was based on a typical pixel screen size. I added a more sophisicated bincount determination here [a10ee20](https://github.com/PhenoApps/Field-Book/pull/985/commits/a10ee205fec25de20b2c4d5b859131e193d34afc), the max is now based on a constant min bin width and on the device metrics.

File: app/src/main/java/com/fieldbook/tracker/adapters/FieldDetailAdapter.kt [hunk 9]
```diff
@@ -25,10 +39,51 @@ class FieldDetailAdapter(private var items: MutableList<FieldDetailItem>) : Recy
 
     override fun onBindViewHolder(holder: ViewHolder, position: Int) {
```
[the remark this answers is not in the document]
Comment (bellerbrock) [reply app/src/main/java/com/fieldbook/tracker/adapters/FieldDetailAdapter.kt:45]: That makes sense, the list of format strings works but certianly doesn't seem ideal for maintaining longer term. 

What do you think about merging it as is, then implementing the approach you have in mind as part of the offbeat improvements?

**Comment by bellerbrock** [comment]
@chaneylc Thanks for your feedback. Should all be adressed now, other than the trait encapsulation question. I also added some additional fixes and improvements, now it's ready for another review!

**Comment by trife** [comment]
- [ ] Use same sort icon on fields list as in field details
- [ ] Sort by most recent first in the fields list for all date choices and auto scroll to the top after sorting (looks like sorting by date may be messed up right now- sending you a db)
- [ ] Move the sync icon to the left of the card
- [ ] Add [icon](https://pictogrammers.com/library/mdi/icon/chart-bar/) for Data card
- [ ] Add [icon](https://pictogrammers.com/library/mdi/icon/land-fields/) for top card
- [ ] Reduce overall `LinearLayout` padding in field details fragment from 16dp to 8dp
- [ ] Modify Data card to be collapsible. Collapsed card shows num traits and obs as [chips](https://m2.material.io/components/chips/android#using-chips) on the same line. Expanded card displays and expands all graphs.
- [ ] Replace narrative in top card with [chips](https://m2.material.io/components/chips/android#using-chips). Add icons to chips to make up for missing words (sort, file, [entries](https://pictogrammers.com/library/mdi/icon/dots-grid/), [attributes](https://pictogrammers.com/library/mdi/icon/information-outline/))

**Comment by trife** [comment]
- [ ] Switch back to appcompat - materialcomponents breaks a lot of the UI (e.g. all dialog boxes now have white text, all buttons in collect are green, etc.). If this means we can't stylize chips right now on a per theme basis, that's fine. But we should probably make an issue to move from appcompat to materialcomponents. Potentially can use a [bridge theme](https://m2.material.io/develop/android/docs/getting-started) to fix this. 
- [ ] Should chips have color (accent or main)?
- [ ] Use a single chip group in top card
- [ ] Hide chip if non-configured/empty (i.e. sort) or zero data (i.e. observations)
- [ ] [Traits icon](https://pictogrammers.com/library/mdi/icon/ruler/)
- [ ] [Observations icon](https://pictogrammers.com/library/mdi/icon/eye/)
- [ ] Thoughts about moving the toolbar features to on chip clicks? Non action clicks could have a toast with a verbose explanation of their info (e.g. 5 field attributes). If we do this we should sort chips to have actions at the end of the list
- [ ] For padding- revert everything back to what it was except the very most outer linearlayout

**Comment by bellerbrock** [comment]
I set the observations chjp to be hidden if there are no observations, but I'm thinking the sort chip should always be visible - brapi fields now require selecting a sort order on import, and its helpful to specify that local fields are ordered the same as the import file if not explictly sorted. Though maybe the file import should also require setting the sort order on import?

I like the idea of moving the sort and rename toolbar tools to on click behavior of the relevant chips. When I see chips I'd expect them to act like buttons.  The trait count chip would then be the logical way to open the field-specific trait activity, and yeah for the other chips would be good for the on click behavior to at least show some additional info. 

I could also see an argument for converting the collect, export, and sync cards into chips within the first card, making use of the chip secondary label to show the date.

**Comment by trife** [comment]
Let's go ahead and move the rename and sort actions to the chip and keep them always visible. Sort can have the text "Import order" if the field hasn't been actively sorted. For many of the files, we don't technically have a way to programmatically sort them to get them to that order (#942). 

Let's keep chips for dialogs and list items for activities that exit the field detail page. Export/Sync both have updates planned that will move them into the latter category so they both get to stay as list items for now (#662, #982). For Data, move the chips under the title but still keep them visible when it's collapsed.

**Comment by bellerbrock** [comment]
> How are traits sorted in Data? Should we sort by percent complete?

They're sorted the same way as in the collect and traits activity. I'm thinking that's probably the expected order.

**Comment by trife** [comment]
Changes based on external review:

- [ ] Let's treat active chips like buttons instead of informational so it's more clear what they do. Fixed text for 'Rename' and 'Sort'
- [ ] For graphs, how difficult is it to add 'Count' and 'Values' labels to axes? Units are currently ambiguous

**Comment by bellerbrock** [comment]
For charts it would be pretty difficult to add axis labels, they are not supported natively in MPAndroidChart and the workarounds with custom textviews are pretty messy. In my opinion I don't think it's worth it, only case I can think of where the axis meanings wouldn't be apparent from the existing labels (a trait with numbered categories and  little data collected)

**Review by trife — CHANGES_REQUESTED** [review]
- [ ] Tutorial choice from appintro not being respected when app is loaded
- [ ] Default collect toolbar icons are currently hidden on install until appearance settings are opened (there was some code in configactivity to fix this at one point)
- [ ] Entries are skipped by default when data has been collected until behavior settings are opened

**Review by trife — APPROVED** [review]

**Pull request merged**